tools/dev-server/dev-server.ts

```diff
@@ -59,10 +59,21 @@ export class DevServer {
    */
   private _bazelMiddleware(req: http.IncomingMessage, res: http.ServerResponse) {
     if (!req.url) {
-      res.end('No url specified. Error');
+      res.statusCode = 500;
+      res.end('Error: No url specified');
       return;
     }
 
+    // Detect if the url escapes the server's root path
+    for (const rootPath of this._rootPaths) {
+      const absoluteRootPath = path.resolve(rootPath);
+      const absoluteJoinedPath = path.resolve(path.posix.join(rootPath, getManifestPath(req.url)));
+      if (!absoluteJoinedPath.startsWith(absoluteRootPath)) {
+        res.statusCode = 500;
+        res.end('Error: Detected directory traversal');
+      }
+    }
+
```

Comment:
Would it be best to just disallow `../` in the requests? If we just want to go with that logic, then we could probably simplify it by doing it using `path.relative`? WDYT?

```ts
for (const rootPath of this._rootPaths) {
  const relativePath = path.relative(rootPath, path.join(rootPath, req.url));
  if (relativePath.startsWith('../')) { throw }
}

```

Reply:
I used the approach recommended by Google's security team. As I see it, it's safer because it ensures that, at the very end, the _resolved_ URL didn't escape the webserver root. 

My first approach wouldn't have caught all cases (https://en.wikipedia.org/wiki/Directory_traversal_attack#Variations_of_directory_traversal)

Reply:
Okay, that sounds reasonable.

I'm unsure about `path.resolve` though. The root paths are not file system paths in the dev-server, but rather Bazel manifest/runfile paths, so resolving paths from the `cwd` is incorrect. It doesn't impact the URL escape check in general, but it's technically not correct. It might be better to use `path.relative` to determine if the paths escapes the root path.